leakcanary-object-watcher-android-androidx/src/main/java/leakcanary/internal/ViewModelClearedWatcher.kt

```diff
@@ -0,0 +1,64 @@
+package leakcanary.internal
+
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.ViewModelProvider
+import androidx.lifecycle.ViewModelProvider.Factory
+import androidx.lifecycle.ViewModelStore
+import androidx.lifecycle.ViewModelStoreOwner
+import leakcanary.AppWatcher.Config
+import leakcanary.ObjectWatcher
+import leakcanary.internal.ViewModelClearedWatcher.Companion.install
+
+/**
+ * [AndroidXFragmentDestroyWatcher] calls [install] to add a spy view model in every
+ * [ViewModelStoreOwner] instance (ie FragmentActivity and Fragment). [ViewModelClearedWatcher]
+ * holds on to the map of view models backing its store. When [ViewModelClearedWatcher] receives
+ * the [onCleared] callback, it adds each live view model from the store to the [ObjectWatcher].
+ */
+internal class ViewModelClearedWatcher(
+  storeOwner: ViewModelStoreOwner,
+  private val objectWatcher: ObjectWatcher,
+  private val configProvider: () -> Config
+) : ViewModel() {
+
+  private val viewModelMap: Map<String, ViewModel>?
+
+  init {
+    // We could call ViewModelStore#keys with a package spy in androidx.lifecycle instead,
+    // however that was added in 2.1.0 and we support AndroidX first stable release. viewmodel-2.0.0
+    // does not have ViewModelStore#keys. All versions currently have the mMap field.
+    viewModelMap = try {
+      val mMapField = ViewModelStore::class.java.getDeclaredField("mMap")
+      mMapField.isAccessible = true
+      @Suppress("UNCHECKED_CAST")
+      mMapField[storeOwner.viewModelStore] as Map<String, ViewModel>
```

Comment:
You could use `ViewModelStore#keys` and catch the no method exception and then go back to the reflection. This way you ensure compatibility with the next versions of viewmodel. This is good for R8 too. 

I'm thinking if a proguard rule to keep the mMap would be appropriate...

Reply:
> This way you ensure compatibility with the next versions of viewmodel.

`keys` is package protected so probably not part of the API and subject to change :).

> I'm thinking if a proguard rule to keep the mMap would be appropriate...

Good point, probably yes